### PR TITLE
Support superclass positions

### PR DESCRIPTION
**Pull request by gbp**
Required for https://github.com/mysociety/verification-pages/issues/314

### CODE REVIEW

File: spec/membership_comparison_spec.rb
Comment (tmtmtmtm) [inline spec/membership_comparison_spec.rb]: Am I reading this correctly as saying that if someone has a blank P39 of "member of the UK Parliament" (for example), the tool will never add a "member of the 44th UK Parliament" statement? If so, then I don't think that's the desired behaviour. I would say it's OK to add a new statement in such a case.

File: lib/membership_comparison/position_comparison.rb
Comment (tmtmtmtm) [inline lib/membership_comparison/position_comparison.rb]: I think there's a typo here, but I'm not sure if the "some" is meant to be a different word.

**Review by tmtmtmtm — COMMENTED** [review]
I'm increasingly finding the tests very difficult to read and reason about (I think it might be time to either abstract those further, or switch to some other approach of making the test cases more human readable), so I'm not entirely sure I'm completely following the logic here, but I think the logic here should be akin to: If there would be a conflict with the _same_ position, then there would still be a conflict here, but if no existing statement would conflict, then we're fine to add; but (in a difference to the existing code), we can't simple "re-use" an existing blank statement if it's not the same Item — we need to create an entire new one instead. 

If that's what this already does, then take this simply as confirmation it's doing the right thing :)

Are you going to handle the reverse case separately (where someone has a membership to a subclass of the position on the Page?

File: spec/membership_comparison_spec.rb
Comment (gbp) [reply spec/membership_comparison_spec.rb]: Yes, you're reading this correctly but we easily can change this. I think we would also need to define how partial matches are classified EG. if a statement for the superclass position has a start date which coincides with the suggestion would this then be a conflict? How about the other more generic party/district qualifiers; if we're changing the 'blank' case then my instinct would be say these aren't conflicts too?

Comment (tmtmtmtm) [reply spec/membership_comparison_spec.rb]: I might be misunderstanding your example, but if there are overlapping dates then I'd say that's absolutely a conflict. If that wouldn't be conflict, what would?

To me there are two key questions in this process:

1: does the suggestion conflict with any existing P39(s)? 
2: (assuming no to the first question): Do we edit an existing P39 to add this information, or do we add a new P39?

For question 1, I don't think it's relevant whether the existing P39 is a superclass, subclass, or the same class — the logic for working out whether there's a clash will be same either way.

For question 2, if there's an existing P39 that matches the suggestion, but we have some additional information, we can add to it; but if the existing statement is for a superclass or subclass position, then we need to flag up that it will need looked at manually (unless we're willing and able to also change the target of the statement, which I'm assuming we're not planning to do yet). 

Where the existing P39 is simply a "bare" claim that the person was a member of a superclass; I don't think that's either a clash that would prevent us from doing anything, *or* a statement that we can reuse: in such a case we're fine to simply create a new P39.

It's certainly possible that I'm overlooking something obvious here, so if there are specific examples (either real or imaginary) you think it's worth considering specifically here, as they'd give different outcomes than you'd expect, then looking at those (and turning them into explicit test cases) definitely seems sensible.

**Review by chrismytton — APPROVED** [review]
I'm finding it hard to fully wrap my brain around this, but it definitely seems like a step in the right direction. I'll leave it up to you and @tmtmtmtm to iron out the details 🙂

File: spec/membership_comparison_spec.rb
Comment (gbp) [reply spec/membership_comparison_spec.rb]: @tmtmtmtm I believe I've covered everything you've suggested in the latest changes. I code has also been refactored some hopefully is a bit more logical to understand. There is still the readability of the specs which we should open a ticket for. But the new specs cover the `bare`, `partial match`, `exact match` and `differing`  which should be all the cases we care about here?

Do you have any remaining questions/ thoughts?

File: spec/membership_comparison_spec.rb
Comment (tmtmtmtm) [inline spec/membership_comparison_spec.rb]: I'm a little worried that this is going to be confusing/misleading to people working with this in the future, or using these tests to understand the code. The speaker of a legislature generally shouldn't be modelled as a subclass of a legislator — it's a distinct position in its own right. Sometimes, however, this _is_ currently modelled like that in Wikidata, so using it as the test case here makes it seem as if we're explicitly testing something about how we handle a member who is also the speaker, which I don't think is what you're doing here. I think it would be much better to make test case explicitly reflect the scenario we're actually working with — i.e. term-specific legislative items.

**Review by tmtmtmtm — COMMENTED** [review]
@gbp other than a comment about the test case appearing to reflect something different to what we're actually trying to solve here, these tests look like a good step to me. 

NB: I'm not sure if this code also handles the reversed case of a suggestion matching an existing P39 to a subclass — if it does I'd say it's worth also adding an explicit set of tests for that, to protect against later regressions. (If it doesn't yet, that's fine too: per my comment on mysociety/verification-pages#548, I think it's OK to handle that separately, as long as we remember that it will also be necessary.)

_Note also: I'm only looking at the test cases here — the details of the implementation are for @chrismytton or someone else to review :)_

File: spec/membership_comparison_spec.rb
Comment (gbp) [reply spec/membership_comparison_spec.rb]: I have updated the items we're using the spec setup in commit 89724b2. Do you think this is enough for now?

**Comment by gbp** [comment]
@tmtmtmtm There are new commits to support subclass positions, the existing subclass statements are being treated the same as the superclass ones, but I'm unsure if this is correct as you've said previously:

> Where the existing P39 is simply a "bare" claim that the person was a member of a **superclass**; I don't think that's either a clash

Should I change this for the subclass case?

File: spec/membership_comparison_spec.rb
Comment (tmtmtmtm) [reply spec/membership_comparison_spec.rb]: I guess this depends on how realistic we want to make the tests, or how much it matters that future readers can make sense of them. I'm finding a lot of them quite confusing now, as they explicitly claim they're relating to a member of the Lok Sabha, but then have seats and parties etc that are Canadian. I think I'd have been more minded to either fake a new position that was "member of the 42nd Canadian Parliament", or the like if we wanted to keep going with the current test data (I don't think the tests actually require these to be real Wikidata items, do they?), or else ignore all the Canadian ones entirely, and have a new setup entirely for the India case.

It's largely your call (or the reviewer's!) as to how much that matters, but my general preference is to make test cases as clear and understandable as possible, as these are, in many ways, the de facto product specifications.

**Comment by tmtmtmtm** [comment]
>> Where the existing P39 is simply a "bare" claim that the person was a member of a superclass; I don't think that's either a clash

> Should I change this for the subclass case?

Hmmm. I fear that this one might be even more complicated, depending on what the subclass actually *is*.

The general principle here is that we want to make sure we're creating *good* P39s. So if the people working with this position have decided that, for example, it should use term-by-term subclasses, then we don't want to be creating memberships to the plain legislative position — we need to follow their lead. 

So where we find a term-specific legislative position being used, I suspect we might want to not only raise a clash for *this* suggestion, but to prevent the creation of *any* P39s to the superclass (to prevent the case we actually hit with India, where we were unaware that the people actively working in that area had already decided to use the term-specific items). This might be one we want to find a way to detect "in the large" at either page creation time, or when loading all the statements, rather than only on a suggestion-by-suggestion basis though, as it largely means the entire Verification Page is probably invalid, and should be re-pointed at a different membership item instead.

However, if this were something that only applied to a subset of members, e.g. a pre-existing "senator for life" ([Q826589](https://www.wikidata.org/wiki/Q826589)) on a page for member of the Italian Senate ([Q13653224](https://www.wikidata.org/wiki/Q13653224)), I think we'd want to raise that as a clash at a single suggestion level (if we're not able to simply amend the existing statement.) We wouldn't want to create a new Q13653224 for them.

If we find something like someone being a Speaker, which is really an entirely separate position (and IMO shouldn't be modelled as a subclass, though it sometimes is), I think it's fine to effectively ignore the existing P39, and proceed with creating the new legislative membership item separately to that.

**Comment by gbp** [comment]
> So where we find a term-specific legislative position being used, I suspect we might want to not only raise a clash for _this_ suggestion, but to prevent the creation of _any_ P39s to the superclass (to prevent the case we actually hit with India, where we were unaware that the people actively working in that area had already decided to use the term-specific items). This might be one we want to find a way to detect "in the large" at either page creation time, or when loading all the statements, rather than only on a suggestion-by-suggestion basis though, as it largely means the entire Verification Page is probably invalid, and should be re-pointed at a different membership item instead.

Ticketed: https://github.com/mysociety/verification-pages/issues/579

> However, if this were something that only applied to a subset of members, e.g. a pre-existing "senator for life" ([Q826589](https://www.wikidata.org/wiki/Q826589)) on a page for member of the Italian Senate ([Q13653224](https://www.wikidata.org/wiki/Q13653224)), I think we'd want to raise that as a clash at a single suggestion level (if we're not able to simply amend the existing statement.) We wouldn't want to create a new Q13653224 for them.

This is what the changes in cc1d611 fixes.

> If we find something like someone being a Speaker, which is really an entirely separate position (and IMO shouldn't be modelled as a subclass, though it sometimes is), I think it's fine to effectively ignore the existing P39, and proceed with creating the new legislative membership item separately to that.

Ticketed: https://github.com/mysociety/verification-pages/issues/580

**Comment by gbp** [comment]
I've ticketed any remaining items where there is still room for improvement. Also there is now a branch which improves the specs readability which I'll rebase once this and #26 is merged https://github.com/everypolitician/membership-comparison/compare/spec-refactor

Unless there are objections I'm going to go ahead and merge. @tmtmtmtm?

**Comment by tmtmtmtm** [comment]
@gbp I fear that this has turned into a bit too much of a beast for me to sensibly be able to comment on: there seems to be too many distinct things happening here for me to really be able to make much sense out of it easily. 

If you were to split this up into a few separate PRs I might be able to look at each in turn, but otherwise I think this one comes down to whether @chrismytton is happy to OK the code in light of the all the previous comments (I think there have been some significant changes since his original thumbs-up).

**Comment by gbp** [comment]
This was merged by accident and reverted. Will open separate PRs to for these features

**Pull request merged**
